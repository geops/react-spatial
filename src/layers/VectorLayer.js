import { unByKey } from 'ol/Observable';
import Layer from './Layer';

/**
 * A class representing vector layer to display on BasicMap
 * @class
 * @inheritDoc
 * @param {Object} [options]
 */
class VectorLayer extends Layer {
  constructor(options = {}) {
    super(options);

    this.hitTolerance = options.hitTolerance || 5;

    // Array of click callbacks
    this.clickCallbacks = [];

    // Add click callback
    if (options.onClick) {
      this.onClick(options.onClick);
    }
  }

  /**
   * Listens to click events on the layer.
   * @param {function} callback Callback function, called with the clicked
   *   features (https://openlayers.org/en/latest/apidoc/module-ol_Feature.html),
   *   the layer instance and the click coordinate.
   */
  onClick(callback) {
    if (typeof callback === 'function') {
      this.clickCallbacks.push(callback);
    } else {
      throw new Error('callback must be of type function.');
    }
  }

  /**
   * Request feature information for a given coordinate.
   * @param {ol.Coordinate} coordinate {@link https://openlayers.org/en/latest/apidoc/module-ol_coordinate.html ol/Coordinate} to request the information at.
   * @returns {Promise<Object>} Promise with features, layer and coordinate
   *  or null if no feature was hit.
   * eslint-disable-next-line class-methods-use-this
   */
  getFeatureInfoAtCoordinate(coordinate) {
    let features = [];

    if (this.map) {
      const pixel = this.map.getPixelFromCoordinate(coordinate);
      features = this.map.getFeaturesAtPixel(pixel, {
        layerFilter: l => l === this.olLayer,
        hitTolerance: this.hitTolerance,
      });
    }

    return Promise.resolve({
      features,
      layer: this,
      coordinate,
    });
  }

  /**
   * Initialize the layer and listen to feature clicks.
   * @param {ol.map} map {@link https://openlayers.org/en/latest/apidoc/module-ol_Map-Map.html ol/Map}
   */
  init(map) {
    super.init(map);

    if (!this.map) {
      return;
    }

    // Listen to click events
    this.singleClickRef = this.map.on('singleclick', e => {
      if (!this.clickCallbacks.length) {
        return;
      }

      this.getFeatureInfoAtCoordinate(e.coordinate)
        .then(d => this.callClickCallbacks(d.features, d.layer, d.coordinate))
        .catch(() => this.callClickCallbacks([], this, e.coordinate));
    });
  }

  /**
   * Call click callbacks with given parameters.
   * This is done in a separate function for being able to modify the response.
   * @private
   */
  callClickCallbacks(features, layer, coordinate) {
    this.clickCallbacks.forEach(c => c(features, layer, coordinate));
  }

  /**
   * Terminate what was initialized in init function. Remove layer, events...
   */
  terminate() {
    super.terminate();
    if (this.singleClickRef) {
      unByKey(this.singleClickRef);
    }
  }
}

export default VectorLayer;
