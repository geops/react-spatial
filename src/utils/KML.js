import KML from 'ol/format/KML';
import { Feature } from 'ol';
import Point from 'ol/geom/Point';
import MultiPoint from 'ol/geom/MultiPoint';
import GeometryCollection from 'ol/geom/GeometryCollection';
import { Style, Text, Icon, Circle, Fill, Stroke } from 'ol/style';
import { asString } from 'ol/color';
import { kmlStyle } from './Styles';
import getPolygonPattern from './getPolygonPattern';

const applyTextStyleForIcon = (olIcon, olText) => {
  const size = olIcon.getSize() || [48, 48];
  const scale = olIcon.getScale() || 1;
  const anchor = olIcon.getAnchor() || [
    (size[0] * scale) / 2,
    (size[1] * scale) / 2,
  ];
  const offset = [
    scale * (size[0] - anchor[0]) + 5,
    scale * (size[1] / 2 - anchor[1]),
  ];
  olText.setOffsetX(offset[0]);
  olText.setOffsetY(offset[1]);
  olText.setTextAlign('left');
};

const getVertexCoord = (geom, start = true, index = 0) => {
  const coords = geom.getCoordinates();
  const len = coords.length - 1;
  return start ? coords[index] : coords[len - index];
};

const getLineIcon = (feature, icon, color, start = true) => {
  const geom = feature.getGeometry();
  const coordA = getVertexCoord(geom, start, 1);
  const coordB = getVertexCoord(geom, start);
  const dx = start ? coordA[0] - coordB[0] : coordB[0] - coordA[0];
  const dy = start ? coordA[1] - coordB[1] : coordB[1] - coordA[1];
  const rotation = Math.atan2(dy, dx);

  return new Style({
    geometry: (feat) => {
      const ge = feat.getGeometry();
      return new Point(getVertexCoord(ge, start));
    },
    image: new Icon({
      src: icon.url,
      color,
      rotation: -rotation,
      rotateWithView: true,
      scale: icon.scale,
      imgSize: icon.size, // ie 11
    }),
    zIndex: icon.zIndex,
  });
};

// Clean the uneeded feature's style and properties created by the KML parser.
const sanitizeFeature = (feature) => {
  const geom = feature.getGeometry();
  let styles = feature.getStyleFunction();

  // The use of clone is part of the scale fix line 156
  const tmpStyles = styles(feature);
  const style = (Array.isArray(tmpStyles) ? tmpStyles[0] : tmpStyles).clone();

  let stroke = style.getStroke();
  if (stroke && feature.get('lineDash')) {
    stroke.setLineDash(
      feature
        .get('lineDash')
        .split(',')
        .map((l) => parseInt(l, 10)),
    );
  }

  // The canvas draws a stroke width=1 by default if width=0, so we
  // remove the stroke style in that case.
  if (stroke && stroke.getWidth() === 0) {
    stroke = undefined;
  }

  if (feature.get('zIndex')) {
    style.setZIndex(parseInt(feature.get('zIndex'), 10));
  }

  // if the feature is a Point and we are offline, we use default vector
  // style.
  // if the feature is a Point and has a name with a text style, we
  // create a correct text style.
  // TODO Handle GeometryCollection displaying name on the first Point
  // geometry.
  if (style && (geom instanceof Point || geom instanceof MultiPoint)) {
    let image = style.getImage();
    let text = null;
    let fill = style.getFill();

    // If the feature has name we display it on the map as Google does
    if (
      feature.get('name') &&
      style.getText() &&
      style.getText().getScale() !== 0
    ) {
      if (image && image.getScale() === 0) {
        // transparentCircle is used to allow selection
        image = new Circle({
          radius: 1,
          fill: new Fill({ color: [0, 0, 0, 0] }),
          stroke: new Stroke({ color: [0, 0, 0, 0] }),
        });
      }

      // We replace empty white spaces used to keep normal spaces before and after the name.
      let name = feature.get('name');
      if (/\u200B/g.test(name)) {
        name = name.replace(/\u200B/g, '');
        feature.set('name', name);
      }

      text = new Text({
        font: feature.get('textFont') || 'normal 16px Helvetica',
        text: feature.get('name'),
        fill: style.getText().getFill(),
        // rotation unsupported by KML, taken instead from custom field.
        rotation: feature.get('textRotation') || 0,
        // since ol 6.3.1 : https://github.com/openlayers/openlayers/pull/10613/files#diff-1883da8b57e690db7ea0c35ce53c880aR925
        // a default textstroke is added to mimic google earth.
        // it was not the case before, the stroke was always null. So to keep
        // the same behavior we don't copy the stroke style.
        // TODO : maybe we should use this functionnality in the futur.
        // stroke: style.getText().getStroke(),
        scale: style.getText().getScale(),
      });

      if (feature.get('textAlign')) {
        text.setTextAlign(feature.get('textAlign'));
      }

      if (feature.get('textOffsetX')) {
        text.setOffsetX(parseFloat(feature.get('textOffsetX')));
      }

      if (feature.get('textOffsetY')) {
        text.setOffsetY(parseFloat(feature.get('textOffsetY')));
      }

      if (feature.get('textBackgroundFillColor')) {
        text.setBackgroundFill(
          new Fill({
            color: feature.get('textBackgroundFillColor'),
          }),
        );
      }

      if (feature.get('textPadding')) {
        text.setPadding(
          feature
            .get('textPadding')
            .split(',')
            .map((n) => parseFloat(n)),
        );
      }

      if (image instanceof Icon) {
        applyTextStyleForIcon(image, text);
      }
    }

    if (image instanceof Icon) {
      /* Apply icon rotation if defined (by default only written as
       * <heading> tag, which is not read as rotation value by the ol KML module)
       */
      image.setRotation(parseFloat(feature.get('iconRotation')) || 0);
    }

    fill = undefined;
    stroke = undefined;

    styles = (feat, resolution) => {
      /* Options to be used for picture scaling with map, should have at least
       * a resolution attribute (this is the map resolution at the zoom level when
       * the picture is created), can take an optional constant for further scale
       * adjustment.
       * e.g. { resolution: 0.123, defaultScale: 1 / 6 }
       */

      if (feat.get('pictureOptions')) {
        let pictureOptions = feat.get('pictureOptions');
        if (typeof pictureOptions === 'string') {
          pictureOptions = JSON.parse(pictureOptions);
        }
        feat.set('pictureOptions', pictureOptions);
        if (pictureOptions.resolution) {
          image.setScale(
            (pictureOptions.resolution / resolution) *
              pictureOptions.defaultScale || 1,
          );
        }
      }

      return new Style({
        fill,
        stroke,
        image,
        text,
        zIndex: style.getZIndex(),
      });
    };

    feature.setStyle(styles);
  }

  // Remove image and text styles for polygons and lines
  if (
    !(
      geom instanceof Point ||
      geom instanceof MultiPoint ||
      geom instanceof GeometryCollection
    )
  ) {
    styles = [
      new Style({
        fill: style.getFill(),
        stroke,
        image: null,
        text: null,
        zIndex: style.getZIndex(),
      }),
    ];

    // Parse the fillPattern json string and store parsed object
    let fillPattern = feature.get('fillPattern');
    if (fillPattern) {
      fillPattern = JSON.parse(fillPattern);
      feature.set('fillPattern', fillPattern);

      /* We set the fill pattern for polygons */
      if (!style.getFill()) {
        styles[0].setFill(new Fill());
      }
      const patternOrColor = fillPattern.empty
        ? [0, 0, 0, 0]
        : getPolygonPattern(fillPattern.id, fillPattern.color);
      styles[0].getFill().setColor(patternOrColor);
    }

    // Add line's icons styles
    if (feature.get('lineStartIcon')) {
      styles.push(
        getLineIcon(
          feature,
          JSON.parse(feature.get('lineStartIcon')),
          stroke.getColor(),
        ),
      );
    }

    if (feature.get('lineEndIcon')) {
      styles.push(
        getLineIcon(
          feature,
          JSON.parse(feature.get('lineEndIcon')),
          stroke.getColor(),
          false,
        ),
      );
    }
    feature.setStyle(styles);
  }
};

/**
 * Read a KML string.
 * @param {String} kmlString A string representing a KML file.
 * @param {<ol.Projection|String>} featureProjection The projection used by the map.
 */
const readFeatures = (kmlString, featureProjection) => {
  const features = new KML().readFeatures(kmlString, {
    featureProjection,
  });
  features.forEach((feature) => {
    sanitizeFeature(feature);
  });
  return features;
};

/**
 * Create a KML string.
 * @param {VectorLayer} layer A react-spatial VectorLayer.
 * @param {<ol.Projection|String>} featureProjection The current projection used by the features.
 */
const writeFeatures = (layer, featureProjection, mapResolution) => {
  let featString;
  const { olLayer } = layer;
  const exportFeatures = [];

  olLayer.getSource().forEachFeature((feature) => {
    // We silently ignore Circle elements as they are
    // not supported in kml.
    if (feature.getGeometry().getType() === 'Circle') {
      return;
    }

    const clone = feature.clone();
    clone.setId(feature.getId());
    clone.getGeometry().setProperties(feature.getGeometry().getProperties());
    clone.getGeometry().transform(featureProjection, 'EPSG:4326');

    // We remove all ExtendedData not related to style.
    Object.keys(feature.getProperties()).forEach((key) => {
      if (!/^(geometry|name|description)$/.test(key)) {
        clone.unset(key, true);
      }
    });

    let styles;

    if (feature.getStyleFunction()) {
      styles = feature.getStyleFunction()(feature, mapResolution);
    } else if (olLayer && olLayer.getStyleFunction()) {
      styles = olLayer.getStyleFunction()(feature, mapResolution);
    }

    const mainStyle = styles[0] || styles;

    const newStyle = {
      fill: mainStyle.getFill(),
      stroke: mainStyle.getStroke(),
      text: mainStyle.getText(),
      image: mainStyle.getImage(),
      zIndex: mainStyle.getZIndex(),
    };

    if (newStyle.zIndex) {
      clone.set('zIndex', newStyle.zIndex);
    }

    // If we see spaces at the beginning or at the end we add a empty
    // white space at the beginning and at the end.
    if (newStyle.text && /^\s|\s$/g.test(newStyle.text.getText())) {
      newStyle.text.setText(`\u200B${newStyle.text.getText()}\u200B`);
    }

    // Set custom properties to be converted in extendedData in KML.
    if (newStyle.text && newStyle.text.getRotation()) {
      clone.set('textRotation', newStyle.text.getRotation());
    }

    if (newStyle.text && newStyle.text.getFont()) {
      clone.set('textFont', newStyle.text.getFont());
    }

    if (newStyle.text && newStyle.text.getTextAlign()) {
      clone.set('textAlign', newStyle.text.getTextAlign());
    }

    if (newStyle.text && newStyle.text.getOffsetX()) {
      clone.set('textOffsetX', newStyle.text.getOffsetX());
    }

    if (newStyle.text && newStyle.text.getOffsetY()) {
      clone.set('textOffsetY', newStyle.text.getOffsetY());
    }

    if (newStyle.text && newStyle.text.getBackgroundFill()) {
      clone.set(
        'textBackgroundFillColor',
        asString(newStyle.text.getBackgroundFill().getColor()),
      );
    }

    if (newStyle.text && newStyle.text.getPadding()) {
      clone.set('textPadding', newStyle.text.getPadding().join());
    }

    if (newStyle.stroke && newStyle.stroke.getLineDash()) {
      clone.set('lineDash', newStyle.stroke.getLineDash().join(','));
    }

    if (newStyle.image instanceof Circle) {
      newStyle.image = null;
    }

    if (newStyle.image) {
      const imgSource = newStyle.image.getSrc();
      if (!/(http(s?)):\/\//gi.test(imgSource)) {
        // eslint-disable-next-line no-console
        console.log(
          'Local image source not supported for KML export.' +
            'Should use remote web server',
        );
      }

      if (newStyle.image.getRotation()) {
        // We set the icon rotation as extended data
        clone.set('iconRotation', newStyle.image.getRotation());
      }

      // Set map resolution to use for icon-to-map proportional scaling
      if (feature.get('pictureOptions')) {
        clone.set(
          'pictureOptions',
          JSON.stringify(feature.get('pictureOptions')),
        );
      }
    }

    // In case a fill pattern should be applied (use fillPattern attribute to store pattern id, color etc)
    if (feature.get('fillPattern')) {
      clone.set('fillPattern', JSON.stringify(feature.get('fillPattern')));
      newStyle.fill = null;
    }

    // If only text is displayed we must specify an
    // image style with scale=0
    if (newStyle.text && !newStyle.image) {
      newStyle.image = new Icon({
        src: 'noimage',
        scale: 0,
      });
    }

    // In case we use line's icon .
    const extraLineStyles = (Array.isArray(styles) && styles.slice(1)) || [];
    extraLineStyles.forEach((extraLineStyle) => {
      if (
        extraLineStyle &&
        extraLineStyle.getImage() instanceof Icon &&
        extraLineStyle.getGeometry()
      ) {
        const coord = extraLineStyle.getGeometry()(feature).getCoordinates();
        const startCoord = feature.getGeometry().getFirstCoordinate();
        if (coord[0] === startCoord[0] && coord[1] === startCoord[1]) {
          clone.set(
            'lineStartIcon',
            JSON.stringify({
              url: extraLineStyle.getImage().getSrc(),
              scale: extraLineStyle.getImage().getScale(),
              size: extraLineStyle.getImage().getSize(),
              zIndex: extraLineStyle.getZIndex(),
            }),
          );
        } else {
          clone.set(
            'lineEndIcon',
            JSON.stringify({
              url: extraLineStyle.getImage().getSrc(),
              scale: extraLineStyle.getImage().getScale(),
              size: extraLineStyle.getImage().getSize(),
              zIndex: extraLineStyle.getZIndex(),
            }),
          );
        }
      }
    });

    const olStyle = new Style(newStyle);
    clone.setStyle(olStyle);

    if (
      !(
        clone.getGeometry() instanceof Point &&
        olStyle.getText() &&
        !olStyle.getText().getText()
      )
    ) {
      exportFeatures.push(clone);
    }
  });

  if (exportFeatures.length > 0) {
    if (exportFeatures.length === 1) {
      // force the add of a <Document> node
      exportFeatures.push(new Feature());
    }

    featString = new KML({
      extractStyles: true,
      defaultStyle: [kmlStyle],
    }).writeFeatures(exportFeatures);

    // Remove no image hack
    featString = featString.replace(
      /<Icon>\s*<href>noimage<\/href>\s*<\/Icon>/g,
      '',
    );

    // Remove empty placemark added to have
    // <Document> tag
    featString = featString.replace(/<Placemark\/>/g, '');

    // Add KML document name
    if (layer.name) {
      featString = featString.replace(
        /<Document>/,
        `<Document><name>${layer.name}</name>`,
      );
    }
  }

  return featString;
};

export default { readFeatures, writeFeatures };
