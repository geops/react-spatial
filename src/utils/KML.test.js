import { configure } from 'enzyme';
import VectorLayer from 'ol/layer/Vector';
import VectorSource from 'ol/source/Vector';
import { get } from 'ol/proj';
import Adapter from 'enzyme-adapter-react-16';
import beautify from 'xml-beautifier';
import KML from './KML';

configure({ adapter: new Adapter() });
const xmlns =
  'xmlns="http://www.opengis.net/kml/2.2" xmlns:gx="http://www.google.com/kml/ext/2.2" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.opengis.net/kml/2.2 https://developers.google.com/kml/schema/kml22gx.xsd"';

const expectWriteResult = (feats, str) => {
  expect(
    beautify(
      KML.writeFeatures(
        {
          name: 'lala',
          olLayer: new VectorLayer({
            source: new VectorSource({
              features: feats,
            }),
          }),
        },
        get('EPSG:4326'),
      ),
    ),
  ).toEqual(beautify(str));
};

describe('KML', () => {
  describe('readFeatures() and writeFeatures()', () => {
    test('should read/write LineStyle and ExtendedData (linesDash, lineStartIcon and lineEndIcon).', () => {
      const str = `
        <kml ${xmlns}>
          <Document>
            <name>lala</name>
            <Placemark>
                <Style>
                  <LineStyle>
                    <color>ff056600</color>
                    <width>1</width>
                  </LineStyle>
                </Style>
                <ExtendedData>
                  <Data name="lineDash"><value>40,40</value></Data>
                  <Data name="lineEndIcon"><value>{"url":"fooarrowend.png","scale":0.35,"size":[36,58]}</value></Data>
                  <Data name="lineStartIcon"><value>{"url":"fooarrowstart.png","scale":0.35,"size":[36,58]}</value></Data>
                </ExtendedData>
                <LineString><coordinates>0,1,0 3,5,0 40,25,0</coordinates></LineString>
            </Placemark>
          </Document>
        </kml>
      `;
      const feats = KML.readFeatures(str);
      const styles = feats[0].getStyle();
      expect(feats.length).toBe(1);
      expect(styles.length).toBe(3);

      // line stroke
      const strokeStyle = styles[0].getStroke();
      expect(strokeStyle.getColor()).toEqual([0, 102, 5, 1]);
      expect(strokeStyle.getLineDash()).toEqual([40, 40]);

      // line start icon
      const lineStartStyle = styles[1];
      expect(lineStartStyle.getImage().getRotation()).toEqual(
        -0.9272952180016122,
      );
      expect(lineStartStyle.getImage().getColor()).toEqual([0, 102, 5, 1]);
      expect(lineStartStyle.getImage().getRotateWithView()).toBe(true);
      expect(lineStartStyle.getGeometry()(feats[0]).getCoordinates()).toEqual([
        0,
        1,
        0,
      ]);

      // line end icon
      const lineEndStyle = styles[2];
      expect(lineEndStyle.getImage().getRotation()).toEqual(
        -0.49555167348582857,
      );
      expect(lineEndStyle.getImage().getRotateWithView()).toBe(true);
      expect(lineEndStyle.getImage().getColor()).toEqual([0, 102, 5, 1]);
      expect(lineEndStyle.getGeometry()(feats[0]).getCoordinates()).toEqual([
        40,
        25,
        0,
      ]);

      expectWriteResult(feats, str);
    });

    test('should read/write TextStyle and ExtendedData.', () => {
      const str = `
        <kml ${xmlns}>
          <Document>
            <name>lala</name>
            <Placemark>
              <name>   bar  </name>
              <Style>
                <IconStyle>
                  <scale>0</scale>
                </IconStyle>
                <LabelStyle>
                  <color>ff7e3420</color>
                  <scale>2</scale>
                </LabelStyle>
              </Style>
              <ExtendedData>
                <Data name="textAlign">
                  <value>right</value>
                </Data>
                <Data name="textBackgroundFillColor">
                  <value>rgba(255,255,255,0.01)</value>
                </Data>
                <Data name="textFont">
                  <value>bold 16px arial</value>
                </Data>
                <Data name="textOffsetX">
                  <value>-90</value>
                </Data>
                <Data name="textOffsetY">
                  <value>30</value>
                </Data>
                <Data name="textPadding">
                  <value>5,6,7,8</value>
                </Data>
                <Data name="textRotation">
                  <value>2.303834612632515</value>
                </Data>
              </ExtendedData>
              <Point>
                <coordinates>0,0,0</coordinates>
              </Point>
            </Placemark>
          </Document>
        </kml>
      `;
      const feats = KML.readFeatures(str);
      const styles = feats[0].getStyle();
      expect(feats.length).toBe(1);
      expect(styles.length).toBe(1);

      // Text
      const style = styles[0].getText();
      expect(style.getText()).toBe('bar'); // spaces are trimmed.
      expect(style.getFont()).toEqual('bold 16px arial');
      expect(style.getFill()).toEqual({ color_: [32, 52, 126, 1] });
      expect(style.getStroke()).toEqual(null);
      expect(style.getScale()).toEqual(2);
      expect(style.getRotation()).toEqual('2.303834612632515');
      expect(style.getPadding()).toEqual([5, 6, 7, 8]);
      expect(style.getBackgroundFill()).toEqual({
        color_: 'rgba(255,255,255,0.01)',
      });
      expect(style.getTextAlign()).toEqual('right');
      expect(style.getOffsetX()).toEqual(-90);
      expect(style.getOffsetY()).toEqual(30);
      expectWriteResult(feats, str);
    });

    test.only('should read and write lineDash and fillPattern style for polygon', () => {
      const str = `
        <kml ${xmlns}>
          <Document>
            <name>lala</name>
            <Placemark>
                <description></description>
                <Style>
                    <LineStyle>
                        <color>ff0000eb</color>
                        <width>2</width>
                    </LineStyle>
                    <PolyStyle>
                      <fill>0</fill>
                    </PolyStyle>
                </Style>
                <ExtendedData>
                    <Data name="fillPattern">
                        <value>{"id":3,"color":[235,0,0,1]}</value>
                    </Data>
                    <Data name="lineDash">
                        <value>1,1</value>
                    </Data>
                </ExtendedData>
                <Polygon>
                    <outerBoundaryIs>
                        <LinearRing>
                            <coordinates>8.521,47.381,0 8.529,47.375,0 8.531,47.382,0 8.521,47.381,0</coordinates>
                        </LinearRing>
                    </outerBoundaryIs>
                </Polygon>
            </Placemark>
        </Document>
      </kml>
      `;
      const feats = KML.readFeatures(str);
      const styles = feats[0].getStyle();
      expect(feats.length).toBe(1);
      expect(styles.length).toBe(1);

      // Polygon
      const feature = feats[0];
      const outlineStyle = styles[0].getStroke();
      expect(outlineStyle.getColor()).toEqual([235, 0, 0, 1]);
      expect(outlineStyle.getWidth()).toEqual(2);
      const fillStyle = styles[0].getFill();
      expect(feature.get('fillPattern')).toEqual({
        id: 3,
        color: [235, 0, 0, 1],
      });
      const color = fillStyle.getColor();
      expect(color.canvas).toMatchSnapshot();
      expect(color.id).toBe(3);
      expect(color.color).toEqual([235, 0, 0, 1]);
      expectWriteResult(feats, str);
    });

    test('should add zIndex and rotation to icon style and zoomAtMaxIconSize to feature properties.', () => {
      const str = `
      <kml ${xmlns}>
        <Document>
            <name>lala</name>
            <Placemark>
                <description></description>
                <Style>
                    <IconStyle>
                        <scale>0.5</scale>
                        <heading>
                          1.5707963267948966
                        </heading>
                        <Icon>
                            <href>https://maps.google.com/mapfiles/kml/pushpin/ylw-pushpin.png</href>
                            <gx:w>64</gx:w>
                            <gx:h>64</gx:h>
                        </Icon>
                        <hotSpot x="20" y="2" xunits="pixels" yunits="pixels"/>
                    </IconStyle>
                </Style>
                <ExtendedData>
                    <Data name="iconRotation">
                      <value>1.5707963267948966</value>
                    </Data>
                    <Data name="zIndex">
                        <value>1</value>
                    </Data>
                    <Data name="zoomAtMaxIconSize">
                      <value>12.65397</value>
                    </Data>
                </ExtendedData>
                <Point>
                    <coordinates>0,0,0</coordinates>
                </Point>
            </Placemark>
        </Document>
      </kml>
      `;
      const feats = KML.readFeatures(str);
      const style = feats[0].getStyle()[0];
      expect(style.getZIndex()).toBe(1);
      expect(style.getImage().getRotation()).toBe(1.5707963267948966);
      expect(feats[0].get('zoomAtMaxIconSize')).toBe(12.65397);
      expectWriteResult(feats, str);
    });
  });
});
