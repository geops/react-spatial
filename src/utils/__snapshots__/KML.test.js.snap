// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KML readFeatures() and writeFeatures() should read and write lineDash and fillPattern style for polygon 1`] = `
<canvas
  data-snapshot-image="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAYAAACNiR0NAAAABmJLR0QA/wD/AP+gvaeTAAAAUElEQVQ4je3UsQmAMBQE0DeJiPuvJMEqSIIrxErQ+gds/g3wuGtO4zjZBHKxNPbGUFkjWGXtlMbolIj1aRZemlhiiSX2Iwbvc5xxtmY1e5be83aXl6guoxkAAAAASUVORK5CYII="
  height="20"
  width="20"
/>
`;
